File: apps/TiflisCode/TiflisCode/Components/Chat/MessageBubble.swift
//
//  MessageBubble.swift
//  TiflisCode
//
//  Copyright (c) 2025 Roman Barinov <rbarinov@gmail.com>
//  Licensed under the Functional Source License (FSL-1.1-NC). See LICENSE file for details.
//

import SwiftUI

/// Chat message bubble with role-based styling
struct MessageBubble: View {
    let message: Message
    let sessionType: Session.SessionType
    var onAction: ((ActionType) -> Void)?

    private var isUser: Bool {
        message.role == .user
    }

    var body: some View {
        HStack(alignment: .top, spacing: 12) {
            if isUser {
                Spacer(minLength: 60)
            } else {
                // Assistant avatar based on session type
                AssistantAvatar(sessionType: sessionType)
            }

            VStack(alignment: isUser ? .trailing : .leading, spacing: 4) {
                // Message content
                MessageBubbleContent(
                    message: message,
                    isUser: isUser,
                    onAction: onAction
                )

                // Timestamp
                Text(message.createdAt, style: .time)
                    .font(.caption2)
                    .foregroundStyle(.tertiary)
            }

            if !isUser {
                Spacer(minLength: 60)
            }
        }
    }
}

/// Avatar for assistant messages using session-specific icon
struct AssistantAvatar: View {
    let sessionType: Session.SessionType

    var body: some View {
        Group {
            if let customIcon = sessionType.customIcon {
                Image(customIcon)
                    .resizable()
                    .aspectRatio(contentMode: .fit)
            } else {
                Image(systemName: sessionType.sfSymbol)
                    .font(.system(size: 18))
                    .foregroundStyle(.primary)
                    .frame(width: 32, height: 32)
                    .background(Color(.systemGray5))
            }
        }
        .frame(width: 32, height: 32)
        .clipShape(RoundedRectangle(cornerRadius: 8))
    }
}

/// Message content with structured content blocks
struct MessageBubbleContent: View {
    let message: Message
    let isUser: Bool
    var onAction: ((ActionType) -> Void)?

    /// Full text content for copying
    private var fullTextContent: String {
        message.contentBlocks.compactMap { block -> String? in
            switch block {
            case .text(_, let text):
                return text
            case .code(_, _, let code):
                return code
            case .thinking(_, let text):
                return text
            case .error(_, let text):
                return text
            case .cancel(_, let text):
                return text
            case .status(_, let text):
                return text
            case .toolCall(_, _, let name, let input, let output, _):
                var parts = [name]
                if let input = input { parts.append("Input: \(input)") }
                if let output = output { parts.append("Output: \(output)") }
                return parts.joined(separator: "\n")
            case .voiceInput(_, _, let transcription, _):
                return transcription
            case .voiceOutput(_, _, let text, _):
                return text
            case .actionButtons:
                return nil
            }
        }.joined(separator: "\n\n")
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            MessageContentView(
                blocks: message.contentBlocks,
                isStreaming: message.isStreaming,
                onAction: onAction
            )
        }
        .padding(12)
        .background(bubbleBackground)
        .clipShape(RoundedRectangle(cornerRadius: 16))
        .contextMenu {
            Button {
                UIPasteboard.general.string = fullTextContent
            } label: {
                Label("Copy", systemImage: "doc.on.doc")
            }

            if isUser {
                Button {
                    onAction?(.sendMessage(fullTextContent))
                } label: {
                    Label("Resend", systemImage: "arrow.clockwise")
                }
            }

            ShareLink(item: fullTextContent) {
                Label("Share", systemImage: "square.and.arrow.up")
            }
        }
    }

    @ViewBuilder
    private var bubbleBackground: some View {
        if isUser {
            Color.accentColor
                .opacity(0.15)
        } else {
            Color(.systemGray6)
        }
    }
}

/// Typing indicator with animated dots
struct TypingIndicator: View {
    let sessionType: Session.SessionType
    @State private var animationPhase = 0.0

    var body: some View {
        HStack(alignment: .top, spacing: 12) {
            AssistantAvatar(sessionType: sessionType)

            HStack(spacing: 4) {
                ForEach(0..<3, id: \.self) { index in
                    Circle()
                        .fill(Color.secondary)
                        .frame(width: 8, height: 8)
                        .scaleEffect(animationScale(for: index))
                        .animation(
                            .easeInOut(duration: 0.5)
                            .repeatForever(autoreverses: true)
                            .delay(Double(index) * 0.15),
                            value: animationPhase
                        )
                }
            }
            .padding(12)
            .background(Color(.systemGray6))
            .clipShape(RoundedRectangle(cornerRadius: 16))

            Spacer(minLength: 60)
        }
        .onAppear {
            animationPhase = 1.0
        }
    }

    private func animationScale(for index: Int) -> CGFloat {
        let phase = animationPhase + Double(index) * 0.3
        return 0.6 + 0.4 * sin(phase * .pi)
    }
}

// MARK: - Preview

#Preview("User Message") {
    MessageBubble(message: .mockUserMessage, sessionType: .supervisor)
        .padding()
}

#Preview("Assistant Message - Supervisor") {
    MessageBubble(message: .mockAssistantMessage, sessionType: .supervisor)
        .padding()
}

#Preview("Assistant Message - Claude") {
    MessageBubble(message: .mockAssistantMessage, sessionType: .claude)
        .padding()
}

#Preview("Assistant Message - Cursor") {
    MessageBubble(message: .mockAssistantMessage, sessionType: .cursor)
        .padding()
}

#Preview("Streaming Message") {
    MessageBubble(message: .mockStreamingMessage, sessionType: .opencode)
        .padding()
}

#Preview("Message with Tool Calls") {
    ScrollView {
        MessageBubble(message: .mockMessageWithToolCalls, sessionType: .claude)
            .padding()
    }
}

#Preview("Message with Error") {
    MessageBubble(message: .mockMessageWithError, sessionType: .supervisor)
        .padding()
}

#Preview("Typing Indicator") {
    VStack(spacing: 20) {
        TypingIndicator(sessionType: .supervisor)
        TypingIndicator(sessionType: .claude)
        TypingIndicator(sessionType: .cursor)
    }
    .padding()
}

File: apps/TiflisCodeAndroid/app/src/main/java/io/tiflis/code/ui/chat/components/MessageBubble.kt
/*
 * Copyright (c) 2025 Roman Barinov <rbarinov@gmail.com>
 * Licensed under the Functional Source License (FSL-1.1-NC). See LICENSE file for details.
 */

package io.tiflis.code.ui.chat.components

import android.content.Intent
import android.widget.TextView
import android.view.MotionEvent
import android.view.ViewConfiguration
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.indication
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.PressInteraction
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ContentCopy
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.Share
import androidx.compose.material.icons.filled.Terminal
import androidx.compose.material3.*
import androidx.compose.material3.ripple
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.toArgb
import androidx.compose.ui.ExperimentalComposeUiApi
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.input.pointer.pointerInteropFilter
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.unit.DpOffset
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.viewinterop.AndroidView
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import io.tiflis.code.R
import io.tiflis.code.domain.models.Message
import io.tiflis.code.domain.models.MessageContentBlock
import io.tiflis.code.domain.models.MessageRole
import io.tiflis.code.domain.models.SessionType
import io.tiflis.code.ui.theme.*
import io.tiflis.code.ui.theme.accentColor
import io.noties.markwon.Markwon

/**
 * A message bubble for displaying chat messages.
 * Mirrors the iOS MessageBubble view with full feature parity.
 * Supports long-press context menu for copy, resend, and share actions.
 */
@OptIn(ExperimentalComposeUiApi::class)
@Composable
fun MessageBubble(
    message: Message,
    sessionType: SessionType,
    onCopyCode: (String) -> Unit,
    onPlayAudioForMessage: ((String) -> Unit)? = null,
    isAudioLoading: (String) -> Boolean = { false },
    isAudioPlaying: (String) -> Boolean = { false },
    onResend: ((String) -> Unit)? = null,
    modifier: Modifier = Modifier
) {
    val isUser = message.role == MessageRole.USER
    val context = LocalContext.current
    val clipboardManager = LocalClipboardManager.current
    var showContextMenu by remember { mutableStateOf(false) }
    var isResending by remember { mutableStateOf(false) }

    // Get full text content for copy/share (includes both Text and VoiceInput transcriptions)
    val textContent = remember(message) {
        message.contentBlocks.mapNotNull { block ->
            when (block) {
                is MessageContentBlock.Text -> block.text
                is MessageContentBlock.VoiceInput -> block.transcription
                else -> null
            }
        }.filter { it.isNotBlank() }.joinToString("\n")
    }

    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = if (isUser) Arrangement.End else Arrangement.Start,
        verticalAlignment = Alignment.Top
    ) {
        // User messages: spacer on left
        if (isUser) {
            Spacer(modifier = Modifier.widthIn(min = 60.dp))
        } else {
            // Assistant messages: avatar on left (like iOS)
            AssistantAvatar(
                sessionType = sessionType,
                modifier = Modifier.size(32.dp)
            )
            Spacer(modifier = Modifier.width(12.dp))
        }

        Box {
            val interactionSource = remember { MutableInteractionSource() }
            val scope = rememberCoroutineScope()
            var longPressJob by remember { mutableStateOf<Job?>(null) }
            var pressOffset by remember { mutableStateOf(Offset.Zero) }
            val longPressTimeoutMs = ViewConfiguration.getLongPressTimeout().toLong()

            Box(
                modifier = Modifier
                    .widthIn(max = 300.dp)
                    .clip(RoundedCornerShape(16.dp))
                    .background(
                        if (isUser) {
                            // iOS uses Color.accentColor.opacity(0.15) for user bubbles
                            MaterialTheme.colorScheme.primary.copy(alpha = 0.15f)
                        } else {
                            // iOS uses Color(.systemGray6) for assistant bubbles
                            MaterialTheme.colorScheme.surfaceContainerHighest
                        }
                    )
                    .indication(interactionSource, ripple())
                    .pointerInteropFilter { event ->
                        when (event.action) {
                            MotionEvent.ACTION_DOWN -> {
                                pressOffset = Offset(event.x, event.y)
                                // Start ripple
                                scope.launch {
                                    interactionSource.emit(PressInteraction.Press(pressOffset))
                                }
                                // Start long press timer
                                longPressJob?.cancel()
                                longPressJob = scope.launch {
                                    delay(longPressTimeoutMs)
                                    showContextMenu = true
                                }
                                true // Consume to ensure we get all subsequent events
                            }
                            MotionEvent.ACTION_UP -> {
                                longPressJob?.cancel()
                                longPressJob = null
                                // End ripple
                                scope.launch {
                                    interactionSource.emit(
                                        PressInteraction.Release(PressInteraction.Press(pressOffset))
                                    )
                                }
                                true
                            }
                            MotionEvent.ACTION_CANCEL -> {
                                longPressJob?.cancel()
                                longPressJob = null
                                // Cancel ripple
                                scope.launch {
                                    interactionSource.emit(
                                        PressInteraction.Cancel(PressInteraction.Press(pressOffset))
                                    )
                                }
                                true
                            }
                            MotionEvent.ACTION_MOVE -> {
                                // Cancel if moved too far (touch slop)
                                val dx = event.x - pressOffset.x
                                val dy = event.y - pressOffset.y
                                val touchSlop = ViewConfiguration.get(context).scaledTouchSlop
                                if (dx * dx + dy * dy > touchSlop * touchSlop) {
                                    longPressJob?.cancel()
                                    longPressJob = null
                                }
                                true
                            }
                            else -> false
                        }
                    }
                    .padding(12.dp)
            ) {
                Column(
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    message.contentBlocks.forEach { block ->
                        MessageContentBlockView(
                            block = block,
                            isUser = isUser,
                            onCopyCode = onCopyCode,
                            onPlayAudioForMessage = onPlayAudioForMessage,
                            isAudioLoading = isAudioLoading,
                            isAudioPlaying = isAudioPlaying
                        )
                    }
                    // Note: Streaming/thinking indicator is now shown as a separate
                    // TypingIndicatorBubble at the bottom of the chat list
                }
            }

            // Context menu dropdown
            DropdownMenu(
                expanded = showContextMenu,
                onDismissRequest = { showContextMenu = false },
                offset = DpOffset(0.dp, 0.dp)
            ) {
                // Copy option
                if (textContent.isNotBlank()) {
                    DropdownMenuItem(
                        text = { Text("Copy") },
                        leadingIcon = {
                            Icon(Icons.Default.ContentCopy, contentDescription = null)
                        },
                        onClick = {
                            clipboardManager.setText(AnnotatedString(textContent))
                            showContextMenu = false
                        }
                    )
                }

                // Resend option (only for user messages)
                if (isUser && textContent.isNotBlank() && onResend != null) {
                    DropdownMenuItem(
                        text = { Text("Resend") },
                        leadingIcon = {
                            Icon(Icons.Default.Refresh, contentDescription = null)
                        },
                        onClick = {
                            // Guard against double-clicks
                            if (isResending) return@DropdownMenuItem
                            isResending = true
                            showContextMenu = false
                            onResend(textContent)
                        }
                    )
                }

                // Share option
                if (textContent.isNotBlank()) {
                    DropdownMenuItem(
                        text = { Text("Share") },
                        leadingIcon = {
                            Icon(Icons.Default.Share, contentDescription = null)
                        },
                        onClick = {
                            val sendIntent = Intent().apply {
                                action = Intent.ACTION_SEND
                                putExtra(Intent.EXTRA_TEXT, textContent)
                                type = "text/plain"
                            }
                            val shareIntent = Intent.createChooser(sendIntent, null)
                            context.startActivity(shareIntent)
                            showContextMenu = false
                        }
                    )
                }
            }
        }

        // Assistant messages: spacer on right
        if (!isUser) {
            Spacer(modifier = Modifier.widthIn(min = 60.dp))
        }
    }
}

/**
 * Avatar for assistant messages using session-specific icon.
 * Mirrors iOS AssistantAvatar view.
 */
@Composable
fun AssistantAvatar(
    sessionType: SessionType,
    modifier: Modifier = Modifier
) {
    val customLogo = sessionType.customLogoRes()

    if (customLogo != null) {
        // Custom logo image (Cursor, Claude, OpenCode, Supervisor)
        Box(
            modifier = modifier.clip(RoundedCornerShape(8.dp)),
            contentAlignment = Alignment.Center
        ) {
            Image(
                painter = painterResource(id = customLogo),
                contentDescription = sessionType.displayName,
                modifier = Modifier.fillMaxSize(),
                contentScale = ContentScale.Fit
            )
        }
    } else {
        // Fallback to Material Icon with colored background (Terminal)
        Box(
            modifier = modifier
                .clip(CircleShape)
                .background(sessionType.accentColor()),
            contentAlignment = Alignment.Center
        ) {
            Icon(
                imageVector = Icons.Default.Terminal,
                contentDescription = sessionType.displayName,
                modifier = Modifier.size(18.dp),
                tint = Color.White
            )
        }
    }
}

/**
 * Get custom logo drawable resource for session type.
 * Returns null for types that should use Material Icon fallback.
 */
private fun SessionType.customLogoRes(): Int? = when (this) {
    SessionType.SUPERVISOR -> R.drawable.ic_tiflis_logo
    SessionType.CURSOR -> R.drawable.ic_cursor_logo
    SessionType.CLAUDE -> R.drawable.ic_claude_logo
    SessionType.OPENCODE -> R.drawable.ic_opencode_logo
    SessionType.TERMINAL -> null // Use Material Icon
}

/**
 * Renders a single content block within a message.
 */
@Composable
fun MessageContentBlockView(
    block: MessageContentBlock,
    isUser: Boolean,
    onCopyCode: (String) -> Unit,
    onPlayAudioForMessage: ((String) -> Unit)? = null,
    isAudioLoading: (String) -> Boolean = { false },
    isAudioPlaying: (String) -> Boolean = { false }
) {
    // Both user and assistant bubbles now have light backgrounds, so use onSurface for text
    val textColor = MaterialTheme.colorScheme.onSurface

    when (block) {
        is MessageContentBlock.Text -> {
            MarkdownText(
                text = block.text,
                textColor = textColor
            )
        }

        is MessageContentBlock.Code -> {
            CodeBlockView(
                code = block.code,
                language = block.language,
                onCopy = { onCopyCode(block.code) }
            )
        }

        is MessageContentBlock.ToolCall -> {
            ToolCallView(
                name = block.name,
                input = block.input,
                output = block.output,
                status = block.status
            )
        }

        is MessageContentBlock.Thinking -> {
            ThinkingView(text = block.text)
        }

        is MessageContentBlock.Status -> {
            StatusView(text = block.text)
        }

        is MessageContentBlock.Error -> {
            ErrorView(text = block.text)
        }

        is MessageContentBlock.VoiceInput -> {
            VoiceInputView(
                transcription = block.transcription,
                durationMs = block.durationMs
            )
        }

        is MessageContentBlock.VoiceOutput -> {
            // Use messageId if present, otherwise fall back to block id
            val audioId = block.messageId ?: block.id
            VoiceOutputView(
                text = block.text,
                durationMs = block.durationMs,
                messageId = audioId,
                isLoading = isAudioLoading(audioId),
                isPlaying = isAudioPlaying(audioId),
                onPlay = {
                    onPlayAudioForMessage?.invoke(audioId)
                }
            )
        }

        is MessageContentBlock.ActionButtons -> {
            ActionButtonsView(
                buttons = block.buttons,
                onButtonClick = { button ->
                    // Handle button action
                    // In a real implementation, this would send an action to the server
                }
            )
        }
    }
}

/**
 * Markdown text renderer using Markwon library.
 * Supports basic markdown formatting: bold, italic, links, lists, etc.
 */
@Composable
fun MarkdownText(
    text: String,
    textColor: Color,
    modifier: Modifier = Modifier
) {
    val context = LocalContext.current
    val markwon = remember { Markwon.create(context) }
    val textColorInt = textColor.toArgb()

    AndroidView(
        modifier = modifier.fillMaxWidth(),
        factory = { ctx ->
            TextView(ctx).apply {
                setTextColor(textColorInt)
                textSize = 14f
                setLineSpacing(0f, 1.2f)
                // Don't consume touch events - let parent Compose handle long press for context menu
                isClickable = false
                isLongClickable = false
                isFocusable = false
                isFocusableInTouchMode = false
                // Explicitly pass through all touch events to parent
                setOnTouchListener { _, _ -> false }
            }
        },
        update = { textView ->
            textView.setTextColor(textColorInt)
            markwon.setMarkdown(textView, text)
        }
    )
}
